lib/dl_zitadel/dl_zitadel/middlewares/flask.py

```diff
@@ -20,20 +24,118 @@
 class FlaskMiddleware:
     _client: clients.ZitadelSyncClient
     _token_storage: services.ZitadelSyncTokenStorage
+    _allow_user_auth: bool = attr.ib(default=False)
+    _allow_service_auth: bool = attr.ib(default=True)
 
     def set_up(self, app: flask.Flask) -> None:
         app.before_request(self.process)
+        app.teardown_request(self.process_response)
+
+    def process_response(self, exception: Optional[BaseException]) -> None:
+        if self._allow_user_auth:
+            flask.response.set_cookie(
+                middlewares_models.ZitadelCookies.ZITADEL_COOKIE, flask_context.get("zitadel_cookie", "")
+            )
+        return None
 
     def process(self) -> flask.Response | None:
+        auth_result = self.auth()
+
+        if auth_result is None:
+            raise werkzeug_exceptions.Unauthorized()
+
+        temp_rci = dl_api_commons_flask_rci.ReqCtxInfoMiddleware.get_temp_rci()
+        dl_api_commons_flask_rci.ReqCtxInfoMiddleware.replace_temp_rci(
+            temp_rci.clone(
+                user_id=auth_result.user_id,
+                user_name=auth_result.user_name,
+                tenant=dl_api_commons_base_models.TenantCommon(),
+                auth_data=auth_result.data,
+            )
+        )
+        flask_context.zitadel_cookies = auth_result.updated_cookies
+
+        return None
+
+    def auth(self) -> middlewares_models.AuthResult | None:
+        result = None
+
+        if result is None and self._allow_user_auth:
+            result = self.auth_user()
+
+        if result is None and self._allow_service_auth:
+            result = self.auth_service()
+
+        return result
+
+    def auth_user(self) -> middlewares_models.AuthResult | None:
+        zitadel_cookie_str = flask.request.cookies.get(middlewares_models.ZitadelCookies.ZITADEL_COOKIE)
+
+        if zitadel_cookie_str is None:
+            LOGGER.info("Zitadel cookie is missing")
+            return None
+        zitadel_cookie = json.loads(base64.b64decode(zitadel_cookie_str))
```

Comment:
let's wrap all the logic about wrapping and unwrapping zitadel cookie into reusable model and not copypaste it